Garden: CIAB hosting dashboard updates

markbiek opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of ARC-1093

## Proposed Changes

This fills in various gaps with the `/ciab` hosting dashboard for garden sites.

* (Part of ARC-1094) Ensure that `Make primary site address` appears for garden sites (when appropriate)
* (Part of ARC-1099) Update the use-your-own-domain flow for garden sites. We skip the plans page and checkout, map the external domain, and continue.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply 194223-ghe-Automattic/wpcom on your sandbox if it hasn't been merged yet and sandbox the public api.
* Use `/tools/garden/` in MC to create a garden site.
* Once the site is provisioned, use the link in the MC tool to visit the wp-admin. For now, this is necessary to ensure that the required capabilities are synced to the Jetpack site.
* Run this PR locally and visit http://calypso.localhost:3000/ciab/sites.
* You should see your garden site in the list.
* Click "Actions" and "Settings". Then click "Overview" on the settings page.
* The domains list should have the "The perfect domain awaits" CTA.
* Click "choose your own", then "Use a domain I already own", then enter a domain name. (I've been using `random-subdomain.biek.dev` but it doesn't really matter since we're not configuring DNS for the domain).
* Click "Continue", then click "Select" under "Connect your domain".
* Your should return to the site overview.

* Click the "Domains" link at the top of the page.
* Click on the "..." Actions menu for the newly mapped domain.
* Click "Make primary site address".
  - Note that there's a 30 minute delay before you can set a new domain to primary. You can skip this by modifying [this line](https://github.com/Automattic/wp-calypso/blob/trunk/client/dashboard/domains/dataviews/actions.tsx#L194) from `! isRecentlyRegistered( item.registration_date )` to `( true || ! isRecentlyRegistered( item.registration_date ) )`. You may also need to comment out line 1244 of `class.wpcom-store-domains-api-endpoints.php` on your sandbox and sandbox the public-api.
* The domain should be made primary

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

markbiek committed:
Fix use-your-own-domain for Garden sites

When connecting an existing domain to a garden site, skip the plans step and checkout. Simply map the domain and continue.

WPprodigy commented on client/dashboard/domains/dataviews/actions.tsx:
```diff
@@ -102,7 +104,14 @@ export const useActions = ( { user, site }: { user: User; site?: Site } ) => {
+					if ( site && isGarden( site ) && isGardenDomain( item ) ) {
+						return false;
+					}
+					return true;
```
What would it take to give our default domains the `DomainSubtype.DEFAULT_ADDRESS`? That feels like it would be best, as I see this field used in quite a few places. And i imagine it would align us well with wpcom platform code as well. Limiting just in the UI here could get us into trouble if a user skips through and hits some endpoints. Ideally we handle this on the server.

WPprodigy commented on client/dashboard/sites/domains/primary-domain-selector.tsx:
```diff
@@ -55,8 +58,8 @@ const PrimaryDomainSelector = ( { domains, site, user }: PrimaryDomainSelectorPr
-			// Additional filtering for wpcom staging domains
-			if ( hasWpcomStagingDomain && domain.wpcom_domain ) {
+			// Additional filtering for wpcom staging domains - exclude default free domains
+			if ( hasWpcomStagingDomain && isDefaultFreeDomain( domain ) ) {
```
Similarly, I think we should make `domain.wpcom_domain` true for garden site default domains. We do own the TLD and need it to be treated uniformly like all the others we own (*.wordpress.com).

WPprodigy commented on client/dashboard/utils/domain.ts:
```diff
@@ -98,6 +99,11 @@ export function shouldUpgradeToMakeDomainPrimary( {
+	// Garden sites don't require plan upgrades to set primary domains
+	if ( isGarden( site ) ) {
+		return false;
+	}
```
I'm okay with doing some of this for testing, but I do believe we're ultimately going to have custom domains be gated by plan (or at least some sort of purchase).

The plans already exist, so it would be worth chasing that down a bit and seeing if we can start testing the real flow by putting those plans on our the garden site first and having the right gates in place.

fushar reviewed (commented):
Oops... we're actually going to remove the site-level domains card in the Overview page 😬 cc: @taipeicoder 

- https://github.com/Automattic/wp-calypso/pull/106499

Let's exclude all relevant logic from this PR?

markbiek committed:
Fix so we can set a domain as primary

arthur791004 commented on client/dashboard/domains/dataviews/actions.tsx:
```diff
@@ -145,11 +145,12 @@ export const useActions = ( { user, site }: { user: User; site?: Site } ) => {
 				supportsBulk: false,
 				callback: ( domains: DomainSummary[] ) => {
+					const domain = domains[ 0 ];
+					const site = sites?.find( ( s ) => s.ID === domain.blog_id );
```
Could we convert the sites into a mapping from ID to site? Otherwise, it’ll perform a linear search on each action for each domain, which could hurt performance if the user has many domains and sites.

markbiek replied:
I can investigate that. Thanks!

arthur791004 commented on client/dashboard/utils/domain.ts:
```diff
@@ -104,7 +104,8 @@ export function shouldUpgradeToMakeDomainPrimary( {
 		!! site.plan?.is_free &&
-		! hasPlanFeature( site, DotcomFeatures.SET_PRIMARY_CUSTOM_DOMAIN )
+		! hasPlanFeature( site, DotcomFeatures.SET_PRIMARY_CUSTOM_DOMAIN ) &&
+		! site.is_garden
```
I'm not sure, but could we give the Garden site the ability to set a primary custom domain through the site plan features? 🤔

markbiek replied:
I think, long term, that's the way to go. But we still don't have a plan setup for CIAB sites yet.

arthur791004 commented on client/dashboard/domains/index.tsx:
```diff
@@ -21,7 +21,8 @@ export function getDomainId( domain: DomainSummary ): string {
 	const { user } = useAuth();
 	const fields = useFields();
-	const actions = useActions( { user } );
+	const { data: sites } = useQuery( sitesQuery() );
```
It would be better to preload sites (without blocking) from the `domainsRoute`

markbiek committed:
Switch to doing site lookups by id instead of using find

markbiek committed:
Preload sites on the domains route

markbiek committed:
Memoize our sites by blog id

markbiek committed:
Fix so we pass an array of sites on sites/domains

markbiek committed:
Revert "Fix use-your-own-domain for Garden sites"

This reverts commit 1c17b7f436d360ebf79556cb5dbd467185b22283.

The redirect needs to happen in the submit method of the domains flow
instead.

markbiek committed:
Allow garden sites to a map a domain from /ciab

For a garden site, bypass the checkout step when mapping a domain.
Simply map the domain and return to the site domain list

arthur791004 reviewed (commented):
Looks good to me 👍
